CMakeLists.txt

```diff
@@ -113,4 +109,10 @@ else()
     )
 endif()
 
+# Build shared library
+add_library(spvgen SHARED ${SPVGEN_SOURCE_FILES})
 target_link_libraries(spvgen glslang HLSL OGLCompiler SPIRV SPIRV-Tools SPIRV-Tools-opt spirv-cross-c khronos_spirv_interface khronos_vulkan_interface)
+
+# Build static library
+add_library(spvgen_static STATIC ${SPVGEN_SOURCE_FILES})
+target_link_libraries(spvgen_static glslang HLSL OGLCompiler SPIRV SPIRV-Tools SPIRV-Tools-opt spirv-cross-c khronos_spirv_interface khronos_vulkan_interface)
```

Comment:
Doesn't this make CMake compile and link everything twice when we want to produce both the .so and the static library? I'd rather avoid this scenario to save build times and reduce the possibility of having mismatched compilation flags or definitions in each target.

Would it be possible to make the shared library link with a static (base) library? I mean something like:
`spvgen --> spvgen_static`
or
```
spvgen_static  ---> spvgen_base
spvgen   ---------------^
```

Reply:
It sounds good, but there must be a source file when building a library. create a dummy file instead? I will try it. Thanks.

Reply:
@kuhar I am trying to make shared library link with a static (base) library, it looks like an empty source file must be created. 

CMakeLists.txt should like:
`
# Build static library
add_library(spvgen_static STATIC ${SPVGEN_SOURCE_FILES})
target_link_libraries(spvgen_static glslang HLSL OGLCompiler SPIRV SPIRV-Tools SPIRV-Tools....)

# Build shared library
add_library(spvgen SHARED "dummy.cpp")
target_link_libraries(spvgen spvgen_static)
`
we can create "dummy.cpp" when it is needed.
=============================================================
or
`
add_library(spvgen_static STATIC ${SPVGEN_SOURCE_FILES})
target_link_libraries(spvgen_static glslang HLSL OGLCompiler SPIRV SPIRV-Tools SPIRV-Tools....)

# Build shared library
add_library(spvgen SHARED ${SPVGEN_SOURCE_FILES})
target_link_libraries(spvgen spvgen_static)
`
This looks very odd. both shared and static contain all source files.
=============================================================

or do you have some other ways? Thanks.

Reply:
@kuhar I am trying to make shared library link with a static (base) library, it looks like an empty source file must be created.

CMakeLists.txt should like:
```
add_library(spvgen_static STATIC ${SPVGEN_SOURCE_FILES})
target_link_libraries(spvgen_static glslang HLSL OGLCompiler SPIRV SPIRV-Tools SPIRV-Tools....)

add_library(spvgen SHARED "dummy.cpp")
target_link_libraries(spvgen spvgen_static)
```
we can create "dummy.cpp" when it is needed.
or
```
add_library(spvgen_static STATIC ${SPVGEN_SOURCE_FILES})
target_link_libraries(spvgen_static glslang HLSL OGLCompiler SPIRV SPIRV-Tools SPIRV-Tools....)

Build shared library
add_library(spvgen SHARED ${SPVGEN_SOURCE_FILES})
target_link_libraries(spvgen spvgen_static)
```
This looks very odd. both shared and static contain all source files.
or do you have some other ways? Thanks.

Reply:
Why do we need both, the static and the shared library?
Is the shared library still in use when we link spvgen statically into amdllpc?

Reply:
I think we need to keep the shared library for use by other AMD-internal tools.

Reply:
@AMD-dwang I looked at the CMake and the code, and I think we may want proceed in a few smaller steps:

1. The cmake file uses a lot of 'global state'. For example, is uses `add_definition`, `include_directories`. This was fine before because there was a single target, but with 2 libraries it'd be better to use per-target functions, e.g., `target_compile_definitions`, `target_include_directories`, etc. Later, we can set it up so that `spvgen_static` makes them `PUBLIC` for `spvgen` (shared). Or alternatively, we can make a base interface library (e.g., `spvgen_base` / `spvgen_lib`) that both `spvgen_static` and `spvgen` link with.

2. It's a great observation that shared libraries require at least one source file. We could use an empty one, but I think it may actually be better to move all the dynamic loading / initialization logic to a source file exclusive to `spvgen` (shared). The static version does not need any of that.

3. Once we have the clean up from 1. and 2. done, we can actually introduce multiple libraries.

IE I propose to split this work into 3 PRs (or more), each getting us closer to the overall goal. How does this sound like to you?

Reply:
@kuhar Thank you for taking the time to read the CMake and the codes.  

> 1. The cmake file uses a lot of 'global state'. For example, is uses `add_definition`, `include_directories`. This was fine before because there was a single target, but with 2 libraries it'd be better to use per-target functions, e.g., `target_compile_definitions`, `target_include_directories`, etc. Later, we can set it up so that `spvgen_static` makes them `PUBLIC` for `spvgen` (shared). Or alternatively, we can make a base interface library (e.g., `spvgen_base` / `spvgen_lib`) that both `spvgen_static` and `spvgen` link with.

I prefer the latter.

> 2. It's a great observation that shared libraries require at least one source file. We could use an empty one, but I think it may actually be better to move all the dynamic loading / initialization logic to a source file exclusive to `spvgen` (shared). The static version does not need any of that.
Correct it, the static libraries also require a source file.  
The dynamic loading / initialization logic doesn't seem to be removed into a source file, we only can create another header file for it.

I will make a version based on you suggestion first.  Thanks again.

Reply:
@kuhar Making a base interface library is useless, static libraries don't actually link, just package all `.o` files. When we want to produce both the .so and the static library, all are only compiled and linked once.

If `spvgen_static` library is used by other projects, we need to provide `glslang HLSL OGLCompiler SPIRV SPIRV-Tools SPIRV-Tools` to be linked, so we are the best to merge all static libraries to `spvgen_static`,  but for current project , it is unnecessary. So this change should be reasonable.

Reply:
@AMD-dwang There a lot to unpack in the last comment and I'm not sure I completely follow.

The way I understand cmake libraries, they provide 2 things: logical project organization and packaging. Generally, want to compile each .cpp file once and use the same compilation flags across the project. Additionally, in the case of spvgen, we have an extra requirement that we want to package it both as a static and shared library.

1. We should make sure that all common source files get compiled exactly once. This is easiest achieved with some base library were we do not care about the exact output format, because other projects won't use it directly. I think this could be a static or an interface library, but I don't know this well enough to tell which one would be better.

2. If we have some functions / code that we want to only got to the shared library or only to the static library, then we need one extra library for each. For example, if we want to have 2 implementations of the `InitSpvGen` function (find .so/dll for `spvgen`, do nothing for `spvgen_static`, we could put it in two separate .cpp files or the same file but guarded with a preprocessor define. This gives one extra library each (static for `spvgen_static` for static, `spvgen` for shared), and base, so 3 libraries total. If we want to ship the exact same code (modulo an empty placeholder source for shared) in both, 2 should be enough.

> If spvgen_static library is used by other projects, we need to provide glslang HLSL OGLCompiler SPIRV SPIRV-Tools SPIRV-Tools to be linked, so we are the best to merge all static libraries to spvgen_static, but for current project , it is unnecessary.  So this change should be reasonable.

I'm not sure I follow this part of the comment. I think what you noted is that because spvgen does not expose its submodules, we don't need to link them as `PUBLIC` and want to make them `PRIVATE` to the static and shared spvgen, right? This sounds correct to me, but does not invalidate points 1. and 2.

Reply:
Even if no project needs to use both the static and the shared version at the same time, I still think making each source file is in exactly one library would make this easier to understand and more maintainable in the long run.

Reply:
> 
> 1. We should make sure that all common source files get compiled exactly once. This is easiest achieved with some base library were we do not care about the exact output format, because other projects won't use it directly. I think this could be a static or an interface library, but I don't know this well enough to tell which one would be better.

We don't use different options to build `glslang HLSL OGLCompiler SPIRV SPIRV-Tools SPIRV-Tools` and we don't hope to change it, because their source codes are the same as github.com, so they are only compiled once. If this is confusing, we can merge them to a big static library(`spvgen_base`).

> 2. If we have some functions / code that we want to only got to the shared library or only to the static library, then we need one extra library for each. For example, if we want to have 2 implementations of the `InitSpvGen` function (find .so/dll for `spvgen`, do nothing for `spvgen_static`, we could put it in two separate .cpp files or the same file but guarded with a preprocessor define. This gives one extra library each (static for `spvgen_static` for static, `spvgen` for shared), and base, so 3 libraries total. If we want to ship the exact same code (modulo an empty placeholder source for shared) in both, 2 should be enough.

I understand what you mean. Adding a extra library means we must change the compilation of other projects, which should be not accepted.  we must at least ensure that other tools that use `spvgen.so\dll` can work without any modification, so I am afraid we can't separate  `spvgen.h`.

Reply:
> We don't use different options to build glslang HLSL OGLCompiler SPIRV SPIRV-Tools SPIRV-Tools and we don't hope to change it, because their source codes are the same as github.com, so they are only compiled once. If this is confusing, we can merge them to a big static library(spvgen_base).

My comment was about spvgen source files, not its dependencies. And yes, a base static library sounds good to me. (Whether it's STATIC or INTERFACE seemed like an implementation detail to me, I'm not pushing for an interface library. Sorry if this caused confusion before.)

> Adding a extra library means we must change the compilation of other projects, which should be not accepted

Why would it change compilation of projects? Even if we add base library with spvgen source that the shared one will link with, we should get exactly the same API surface.

> we must at least ensure that other tools that use spvgen.so\dll can work without any modification, so I am afraid we can't separate spvgen.h.

The best way to ensure things like this is by adding tests. We can have unit tests that load spvgen dll/so and make sure it works, and a separate set of tests linking with the static library.
As for separating spvgen.h, I don't see why we couldn't factor out common stuff to a new header like `spvgen_base` and use it with projects that link statically, and then include it in spvgen.h so that you end up with the same contents as currently  .

I'm not saying we need to do all of this immediately or in one step, but I want to make sure future modifications and maintenance will be easy. That's why I'm suggesting to stick to the standard ways of organizing cmake projects.

Reply:
If you put `InitSpvGen` function into a `.cpp`, how we call this function before loading dynamic library?

Reply:
> If you put InitSpvGen function into a .cpp, how we call this function before loading dynamic library?

I was thinking of having  two implementations that we switch on based on a define or project files.

1. Option 1 with a define:

In cmake:
```
target_compile_definitions(spvgen_static PUBLIC SPVGEN_DYNAMIC_INIT=0)
target_compile_definitions(spvgen PUBLIC SPVGEN_DYNAMIC_INIT=1)
```

In the headers
```
// In spvgen_common.h (or spvgen.h directly)
void InitSpvGen(const char* path);

// In spvgen.h
#if SPVGEN_DYNAMIC_INIT == 1
void InitSpvGen(const char* path) { InitSpvGenShared(path); } // Where InitSpvGenShared is the existing code.

#else
void InitSpvGen(const char* path) { } // No-op or call an implementation in a cpp files specific to the static version.
#endif
```

2. Option 2 with different source files:

In cmake
```
// Have 2 directories, src/static and src/shared. Both contain spvgen_init.h.

target_include_directories(spvgen_static PUBLIC src/static)
target_include_directories(spvgen PUBLIC src/shared)
```

In spvgen.h:
```
#include "spvgen_init.h"
```

Reply:
Oh,  this doesn't seem to work. we are using `loadLibary` load `spvgen.dll`. Before loading `spvgen.so`, we should not call the function of `spvgen.so`. I need to try it. Thanks.